File: src/stack/c/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "stack.h"

struct Stack* createStack(){
    struct Stack *stack = malloc(sizeof(struct Stack));
    stack->pointer = -1;

    return stack;
}

int push(struct Stack *stack, int value){
    if (isFull(stack)){
        printf("Stack is full!\n");
        return -1;
    }

    stack->pointer++;
    stack->values[stack->pointer] = value;

    return 0;
}

int pop(struct Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    int value = stack->values[stack->pointer];
    stack->pointer--;

    return value;
}

int peek(struct Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    return stack->values[stack->pointer];
}

int print(struct Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    for (int i=stack->pointer; i>=0; i--){
        printf("[ %d ]\n", stack->values[i]);
    }

    return 0;
}

int isFull(struct Stack *stack){
    return stack->pointer == STACK_SIZE - 1;
}

int isEmpty(struct Stack *stack){
    return stack->pointer == -1;
}

File: src/stack/ts/stack.ts
import { StackProps } from "./types"

export class Stack {
    private props: StackProps

    constructor(stackSize: number) {
        if (typeof stackSize !== "number") {
            throw new Error("The size of the stack must be a number")
        } else if (stackSize <= 0) {
            throw new Error("The size of the stack must be greater than 0")
        }

        this.props = {
            stackSize,
            values: [],
            stackPointer: -1
        }
    }

    public push(value: number): void {
        if (typeof value !== "number") {
            throw new Error("Value must be a number")
        } else if (this.props.stackPointer === this.props.stackSize - 1) {
            throw new Error("Stack is full")
        }

        this.props.stackPointer++
        this.props.values[this.props.stackPointer] = value
    }

    public pop(): number {
        if (this.props.stackPointer === -1) {
            throw new Error("Stack is empty")
        }

        const poppedValue = this.props.values[this.props.stackPointer]
        this.props.values.length -= 1
        this.props.stackPointer -= 1

        return poppedValue
    }

    public peek(): number {
        return this.props.values[this.props.stackPointer]
    }

    public isEmpty(): boolean {
        return this.props.stackPointer === -1
    }

    public isFull(): boolean {
        return this.props.stackPointer === this.props.stackSize - 1
    }

    public print(): void {
        if (this.isEmpty()) {
            throw new Error("Stack is empty")
        }

        let output = ""
        for (let i = this.props.values.length - 1; i >= 0; i--) {
            output += `[ ${this.props.values[i]} ]\n`
        }

        console.log(output)
    }
}
